### ISSUE TITLE
N1curl elements don't appear to work correctly

### ISSUE THREAD

**Issue by pefarrell** [firedrakeproject/firedrake#1625]
Consider the following FEniCS and firedrake codes:

FEniCS:
```
from dolfin import *
import argparse

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument("--k", type=int, default=1)
parser.add_argument("--nref", type=int, default=0)
args, _ = parser.parse_known_args()

baseN = 4

k = Constant(args.k)
nref = args.nref

mesh = UnitCubeMesh(baseN*2**nref, baseN*2**nref, baseN*2**nref)

(x, y, z) = SpatialCoordinate(mesh)

#B_ex = as_vector([cos(z)*sin(x), sin(x), exp(z)*x])
B_ex = Expression(("cos(x[2])*sin(x[0])","sin(x[0])","exp(x[2])*x[0]"), degree=4, domain=mesh)

Z = FunctionSpace(mesh, "N1curl", args.k) 

B = TrialFunction(Z)
C = TestFunction(Z)
f =  B_ex + curl(curl(B_ex))
F = (
      inner(B, C)*dx
    + inner(curl(B), curl(C))*dx
     )

L = inner(f, C) * dx

B = Function(Z)
bcs = DirichletBC(Z, B_ex, "on_boundary")

solve(F == L, B, bcs)

error_l2 = errornorm(B_ex, B, 'L2')
error_hcurl = errornorm(B_ex, B, 'hcurl')

print("nref = %s" % nref)
print("||B_ex - B||_L^2 = %s" % error_l2)
print("||B_ex - B||_Hcurl = %s" % error_hcurl)
```

Firedrake:
```
from firedrake import *
import argparse

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument("--k", type=int, default=1)
parser.add_argument("--nref", type=int, default=0)
args, _ = parser.parse_known_args()

baseN = 4

k = Constant(args.k)
nref = args.nref

mesh = UnitCubeMesh(baseN*2**nref, baseN*2**nref, baseN*2**nref)

(x, y, z) = SpatialCoordinate(mesh)

B_ex = as_vector([cos(z)*sin(x), sin(x), exp(z)*x])
#B_ex = Expression(("cos(x[2])*sin(x[0])","sin(x[0])","exp(x[2])*x[0]"), degree=4, domain=mesh)

Z = FunctionSpace(mesh, "N1curl", args.k) 

B = TrialFunction(Z)
C = TestFunction(Z)
f =  B_ex + curl(curl(B_ex))
F = (
      inner(B, C)*dx
    + inner(curl(B), curl(C))*dx
     )

L = inner(f, C) * dx

B = Function(Z)
bcs = DirichletBC(Z, B_ex, "on_boundary")

solver_parameters = {"ksp_type": "preonly", "pc_type": "lu", "pc_factor_mat_solver_type": "mumps"}
solve(F == L, B, bcs, solver_parameters=solver_parameters)

error_l2 = errornorm(B_ex, B, 'L2')
error_hcurl = errornorm(B_ex, B, 'hcurl')

print("nref = %s" % nref)
print("||B_ex - B||_L^2 = %s" % error_l2)
print("||B_ex - B||_Hcurl = %s" % error_hcurl)
```

Running `for nref in 0 1 2; do python3 curl_fenics.py --nref $nref; done` yields
```
nref = 0
||B_ex - B||_L^2 = 0.20823858176363025
||B_ex - B||_Hcurl = 0.25384243655984984
Solving linear variational problem.
nref = 1
||B_ex - B||_L^2 = 0.10430484722284414
||B_ex - B||_Hcurl = 0.12681941189832838
Solving linear variational problem.
nref = 2
||B_ex - B||_L^2 = 0.0521753507750424
||B_ex - B||_Hcurl = 0.06336116986689634
```
i.e. clean first-order convergence in both norms, as expected from the theory (chap 3 of the FEniCS book).

Unfortunately, running `for nref in 0 1 2; do python3 curl_firedrake.py --nref $nref; done` does not yield first-order convergence in the H(curl) norm:
```
nref = 0
||B_ex - B||_L^2 = 0.19919699461723045
||B_ex - B||_Hcurl = 0.7413338476624576
nref = 1
||B_ex - B||_L^2 = 0.11206270581942429
||B_ex - B||_Hcurl = 0.505296802802699
nref = 2
||B_ex - B||_L^2 = 0.05952920509381887
||B_ex - B||_Hcurl = 0.332187102717233
```

Why is the same code (essentially) giving the right answer in FEniCS, but the wrong answer in Firedrake?

**Comment by dham** [comment]
Given the near-identical L2 convergence, my first question is whether our H(curl) error norm is bugged. I bet it's not often used. Do you also see this issue if you code your own error norm form?

**Comment by wence-** [comment]
I think it's overintegration of the nonpolynomial thing, If I write
```
v = B_ex - B
error_hcurl = assemble((inner(v, v) + inner(curl(v), curl(v)))*dx(degree=2))
```

Then I get:
```
nref = 0
||B_ex - B||_L^2 = 0.19919699461723045
||B_ex - B||_Hcurl = 0.5495284158237274
nref = 1
||B_ex - B||_L^2 = 0.11206270581942435
||B_ex - B||_Hcurl = 0.25532181276350957
nref = 2
||B_ex - B||_L^2 = 0.05952920509381887
||B_ex - B||_Hcurl = 0.11034807801332565
```

**Comment by FabianL1908** [comment]
I think the error norm is not the problem. Lawrence, there is a sqrt missing in your formula. If I add the sqrt then your errornorm gives the same results as before.

**Comment by wence-** [comment]
Oh yeah. Does the problem disappear with `export PYOP2_DEBUG=1`?

**Comment by wence-** [comment]
Nope. That doesnt do anything. 

Works fine for N2curl  fwiw

**Comment by FabianL1908** [comment]
I think there is a problem with the boundary conditions here. If I choose a test problem that has homogeneous boundary conditions I can observe clean first order. For example:
```
hom = x*(1-x)*y*(1-y)*z*(1-z)
B_ex = as_vector([hom, hom, hom])
```
with 
`bcs = DirichletBC(Z, 0, "on_boundary")` 
converges with clean first order.

Surprisingly changing the last line to 
`bcs = DirichletBC(Z, B_ex, "on_boundary")`
makes the order worse again.

**Comment by wence-** [comment]
I think what is going on is the following.

Dolfin interpolates the expression to the boundary data, whereas firedrake projects the expression to the boundary data. For N1curl elements, the lowest degree option does not span all the linear polynomials, the optimal L2 projection of the boundary data is therefore only O(h^{1/2}) accurate.

For example, if I run this:
```python
import argparse

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument("--k", type=int, default=1)
parser.add_argument("--nref", type=int, default=0)
parser.add_argument("--dolfin", action="store_true", default=False)
args, _ = parser.parse_known_args()

if args.dolfin:
    from dolfin import *
else:
    from firedrake import *

baseN = 4

k = Constant(args.k)
nref = args.nref

mesh = UnitCubeMesh(baseN*2**nref, baseN*2**nref, baseN*2**nref)

(x, y, z) = SpatialCoordinate(mesh)

B_ex = as_vector([x, 0, 0])

Z = FunctionSpace(mesh, "N1curl", args.k) 


if args.dolfin:
    B_exf = project(B_ex, Z, solver_type="cg", preconditioner_type="icc")
else:
    B_exf = project(B_ex, Z)
if args.dolfin:
    B_ex = Expression(("x[0]", "0", "0"), degree=4, domain=mesh)
    B_exint = interpolate(B_ex, Z)
    print("Interpolation error: ", errornorm(B_ex, B_exint, norm_type="hcurl", degree_rise=1))

print("Projection error: ", errornorm(B_ex, B_exf, norm_type="hcurl", degree_rise=1))
```

I see the same "bad" behaviour in both cases.

e.g. for dolfin:
```
fenics@1f41cc778acf:/code$ python3 bork-curl.py --dolfin --nref 0 --k 1
Interpolation error:  0.07216878364870323
Projection error:  0.318921392098485
fenics@1f41cc778acf:/code$ python3 bork-curl.py --dolfin --nref 1 --k 1
Interpolation error:  0.03608439182435108
Projection error:  0.25515568177111997
fenics@1f41cc778acf:/code$ python3 bork-curl.py --dolfin --nref 2 --k 1
Interpolation error:  0.01804219591217542
Projection error:  0.19099277303567289
```

If we could interpolate into N1curl spaces, then we could do the same thing as dolfin does in its bc.

**Comment by pefarrell** [comment]
OK, that clears things up. What's stopping us from interpolating into N1curl spaces? I've heard before something about the lack of dual element evaluation in FIAT, but DOLFIN seems to manage it.

**Comment by wence-** [comment]
> What's stopping us from interpolating into N1curl spaces? I've heard before something about the lack of dual element evaluation in FIAT, but DOLFIN seems to manage it.

FIAT does have an "interface" for dual evaluation, but it's not great for our purposes: you can ask the element for its dual basis and get the quad rules that implements the functionals. The thing stopping us is that we haven't implemented/wrapped it nicely in finat and wrapped it up in tsfc (both David and I have proposed it a number of times as student projects in various contexts but it's not been picked).

**Comment by wence-** [comment]
We fixed this in #1650.
